python_modules/dagster/dagster/_core/decorator_utils.py

```diff
@@ -128,3 +147,151 @@ def has_at_least_one_parameter(
+
+
+def apply_pre_call_decorator(obj: T_Decoratable, pre_call_fn: Callable[[], None]) -> T_Decoratable:
+    target = get_decorator_target(obj)
+    new_fn = _wrap_with_pre_call_fn(target, pre_call_fn)
+    return _update_decoratable(obj, new_fn)
+
+
+def _wrap_with_pre_call_fn(
+    fn: T_Callable,
+    pre_call_fn: Callable[[], None],
+) -> T_Callable:
+    @functools.wraps(fn)
+    def new_fn(*args, **kwargs):
+        pre_call_fn()
+        return fn(*args, **kwargs)
+
+    return cast(T_Callable, new_fn)
+
+
+def apply_context_manager_decorator(
+    obj: T_Decoratable, cm: Callable[[], ContextManager[None]]
+) -> T_Decoratable:
+    target = get_decorator_target(obj)
+    new_fn = _wrap_with_context_manager(target, cm)
+    return _update_decoratable(obj, new_fn)
+
+
+def _wrap_with_context_manager(
+    fn: T_Callable,
+    cm: Callable[[], ContextManager[None]],
+) -> T_Callable:
+    @functools.wraps(fn)
+    def fn_with_warning(*args, **kwargs):
+        with cm():
+            return fn(*args, **kwargs)
+
+    return cast(T_Callable, fn_with_warning)
+
+
+def _update_decoratable(decoratable: T_Decoratable, new_fn: Callable) -> T_Decoratable:
+    """Update a property with a new `fget` function.
+
+    `property` objects are immutable, so if we want to apply a decorator to the underlying `fget`,
+    we need to create a new `property` object with the updated `fget` function that preserves the
+    rest of the property's attributes. We also want to make sure we copy over any special attributes
+    we stored on the `fget` (since we can't and therefore don't store them on the `property`
+     object).
+    """
+    from dagster._annotations import copy_annotations
+
+    curr_fn = get_decorator_target(decoratable)
+    new_fn = functools.update_wrapper(new_fn, curr_fn)
+    # from types import FunctionType
+    # if inspect.isfunction(decoratable) or isinstance(decoratable, FunctionType):
+    #     val = new_fn
+    if isinstance(decoratable, property):
+        val = property(new_fn, decoratable.fset, decoratable.fdel, decoratable.__doc__)
+        copy_annotations(val, decoratable)
+    elif isinstance(decoratable, staticmethod):
+        val = staticmethod(new_fn)
+        copy_annotations(val, decoratable)
+    elif isinstance(decoratable, classmethod):
+        val = classmethod(new_fn)
+        copy_annotations(val, decoratable)
+    elif isinstance(decoratable, type):
+        constructor_name = "__new__" if issubclass(decoratable, tuple) else "__init__"
+        setattr(decoratable, constructor_name, new_fn)
+        val = decoratable
+    elif is_resource_def(decoratable):
+        setattr(decoratable, "_resource_fn", new_fn)
+        val = decoratable
+    elif is_op_def(decoratable):
+        if is_decorated_op_function(decoratable.compute_fn):
+            setattr(decoratable.compute_fn, "decorated_fn", new_fn)
+        else:
+            setattr(decoratable, "compute_fn", new_fn)
+        val = decoratable
```

Comment:
yikes, why do we need special casing for op and resource?

Reply:
Because these are objects that store the function they wrap under a special attribute. In this they are similar to `property` (which uses `fget`, and `staticmethod`/`classmethod` (which use `__func__`). If you want to be able to do e.g. this:

```
@deprecated(breaking_version="0.21", additional_warn_text="Use DbtCli instead.")
@dagster_maintained_resource
@resource(config_schema=DbtCliClientResource.to_config_schema())
def dbt_cli_resource(context) -> DbtCliClient:
    """This resource issues dbt CLI commands against a configured dbt project."""
```

You need to know how to get at the function the `ResourceDefinition` is wrapping.

Reply:
are these covered in the test suite at all?

Reply:
Yes, see `test_deprecated_resource` and `test_experimental_resource`.